### ISSUE TITLE
Reload config does not work in 2.3.0

### ISSUE THREAD

**Issue by suyograo** [elastic/logstash#4862]
Top of the branch in 2.3, config reloading does not work as expected

```
suyog@machine:/tmp/logstash-2.3.0.snapshot3$ bin/logstash -f ~/ws/elastic/conf/complete.conf --auto-reload
Settings: Default pipeline workers: 8
Pipeline main started
183.60.215.50 - - [01/Jun/2015:18:00:00 +0000] "GET /scripts/netcat-webserver HTTP/1.1" 200 182 "-" "Mozilla/5.0 (compatible; EasouSpider; +http://www.easou.com/search/spider.html)"
.fetched new config for pipeline. upgrading.. {:pipeline=>"main", :config=>"input {\n  stdin { }\n}\n\nfilter {\n  grok {\n    match => {\n      \"message\" => '%{IPORHOST:clientip} %{USER:ident} %{USER:auth} \\[%{HTTPDATE:timestamp}\\] \"%{WORD:verb} %{DATA:request} HTTP/%{NUMBER:httpversion}\" %{NUMBER:response:int} (?:-|%{NUMBER:bytes:int}) %{QS:referrer} %{QS:agent}'\n    }\n  }\n\n  date {\n    match => [ \"timestamp\", \"dd/MMM/YYYY:HH:mm:ss Z\" ]\n    locale => en\n  }\n\n  geoip {\n    source => \"clientip\"\n  }\n\n  useragent {\n    source => \"agent\"\n    target => \"useragent\"\n  }\n}\n\noutput {\n  stdout { codec => rubydebug }\n  elasticsearch {\n    hosts => \"localhost\"\n  }\n}\n\n"}
stopping pipeline {:id=>"main"}
183.60.215.50 - - [01/Jun/2015:18:00:00 +0000] "GET /scripts/netcat-webserver HTTP/1.1" 200 182 "-" "Mozilla/5.0 (compatible; EasouSpider; +http://www.easou.com/search/spider.html)"
Pipeline main has been shutdown
Settings: Default pipeline workers: 8
Pipeline main started
Pipeline main has been shutdown
```

Pretty simple config

```
input {
  stdin { }
}

filter {
  grok {
    match => {
      "message" => '%{IPORHOST:clientip} %{USER:ident} %{USER:auth} \[%{HTTPDATE:timestamp}\] "%{WORD:verb} %{DATA:request} HTTP/%{NUMBER:httpversion}" %{NUMBER:response:int} (?:-|%{NUMBER:bytes:int}) %{QS:referrer} %{QS:agent}'
    }
  }

  date {
    match => [ "timestamp", "dd/MMM/YYYY:HH:mm:ss Z" ]
    locale => en
  }

  geoip {
    source => "clientip"
  }

  useragent {
    source => "agent"
    target => "useragent"
  }
}

output {
  stdout { codec => rubydebug }
  elasticsearch {
    hosts => "localhost"
  }
}
```

Changed `stdout { codec => rubydebug }` to `stdout { codec => dots }` and I enter a new input, it blows

**Comment by suyograo** [comment]
So its the `stdout` output, which I remember you telling me it does not work. I remember now, but its just bad behavior as far as the user is concerned (if they are not expecting it)

Would it be possible to warn or not start when `--auto-reload` is used with streams like `stdin` which cannot be reliably closed and restarted? We can keep a list of such plugins internally and refuse to start LS when they are used in conjunction with auto-reload. We may dealing with a lot of same bug reports otherwise.

**Comment by jsvd** [comment]
@suyograo WDYT:

```
% bin/logstash -f config  -r
unable to load configuration due to plugins incompatible with reloading {:plugins=>[LogStash::Inputs::Stdin], :level=>:error}
% bin/logstash -f config 
Settings: Default pipeline workers: 4
Pipeline main started
SIGHUP received. {:level=>:warn}
unable to reload configuration due to plugins incompatible with reloading {:plugins=>[LogStash::Inputs::Stdin], :level=>:error}
```

**Comment by suyograo** [comment]
@jsvd how about:

```
Logstash could not be started because -r or --auto-reload flag was used with a configuration that is incompatible to dynamic reloading. {:plugins=>[LogStash::Inputs::Stdin], :level=>:error}
```

**Comment by jsvd** [comment]
+1  for the auto_reloading case. Any suggestion for the scenario where -r is not used (so the pipeline starts), and then the config is changed to use stdin{} and SIGHUP is sent?

**Comment by suyograo** [comment]
WDYT of this:

```
Signal received to reload configuration, but unable to do so because of plugins that do not support dynamic reloading {:plugins=>[LogStash::Inputs::Stdin], :level=>:error}
```

**Comment by suyograo** [comment]
Fixed in #4884
